hashintel/hash
# Fixing Distributed Tasks

## 🌟 What is the purpose of this PR?

We allow packages to distribute tasks across workers. Within that we also automatically distribute groups of agents.
Previously we were waiting on tasks to complete, but we weren't waiting on the requests relating to individual groups to complete. This meant that when one group's execution completed, the Engine would continue to the next steps, thus breaking some of our restrictions about mutability on batches.

This also **possibly** fixes a known issue where sometimes a simulation run would fail due to failing to acquire a write lock on batches (similar to that described #263)

## 🐾 Next steps

- Working through this highlighted some improvements we wish to make around how we control read/write access to the state objects. At some point soon we will try and simplify the designs and make them more consistent.
- Implement these changes for the Python runner

## 🔍 What does this change?
- Comments out the cancel logic, we already were mostly ignoring it, this just commits to it and adds TODOs
- Adds `group_index` to the TaskMsg Table in the Runner Outbound Message Flatbuffers schema
- Fixes the genfbs script and regenerates the flatbuffers files
- Adds `PendingGroups` to the `PendingTask` struct, only marking a task as finished when all groups are finished.

### Does this require a change to the docs?

- No, except the small change made to the flatbuffers bit

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201707629991362/1201764677068619/f) (_internal_)
- [Asana task for the write lock bug](https://app.asana.com/0/1201481007343159/1201545519802517/f)(_internal_)

## 🛡 What tests cover this?

- At the moment, none really as it seems to only happen at large agent sizes, but this shouldn't break current integration tests (and it _perhaps_ fixes errors we have been seeing in the tests up until now)

## ❓ How to test this?



1.  Checkout the branch / view the deployment
2. Try running a simple simulation like age, with >200,000 agents made in a simple init.js file
3. Should run without error
